gattc_nina.go

```diff
@@ -0,0 +1,111 @@
+//go:build nina || nano_rp2040
+
+package bluetooth
+
+type uuidWrapper = UUID
+
+// DeviceService is a BLE service on a connected peripheral device.
+type DeviceService struct {
+	uuidWrapper
+
+	device *Device
+}
+
+// UUID returns the UUID for this DeviceService.
+func (s *DeviceService) UUID() UUID {
+	return s.UUID()
+}
+
+// DiscoverServices starts a service discovery procedure. Pass a list of service
+// UUIDs you are interested in to this function. Either a slice of all services
+// is returned (of the same length as the requested UUIDs and in the same
+// order), or if some services could not be discovered an error is returned.
+//
+// Passing a nil slice of UUIDs will return a complete list of
+// services.
+func (d *Device) DiscoverServices(uuids []UUID) ([]DeviceService, error) {
+	d.services = []DeviceService{}
+
+	startHandle := uint16(0x0001)
+	endHandle := uint16(0xffff)
+	for endHandle == uint16(0xffff) {
+		err := d.adapter.att.readByGroupReq(d.handle, startHandle, endHandle, GATT_SERVICE_UUID)
+		if err != nil {
+			return nil, err
+		}
+
+		if len(d.adapter.att.services) > 0 {
+			for _, rawService := range d.adapter.att.services {
+				d.services = append(d.services,
+					DeviceService{
+						uuidWrapper: NewUUID(rawService.uuid),
+					})
+				startHandle = rawService.endHandle + 1
+				if startHandle == 0x0000 {
+					endHandle = 0x0000
+				}
+			}
+
+			// reset raw services
+			d.adapter.att.services = []rawService{}
+		} else {
+			break
+		}
+	}
+
+	return d.services, nil
+}
+
+// DeviceCharacteristic is a BLE characteristic on a connected peripheral
+// device.
+type DeviceCharacteristic struct {
+	uuid shortUUID
+
+	service *DeviceService
+}
+
+// UUID returns the UUID for this DeviceCharacteristic.
+func (c *DeviceCharacteristic) UUID() UUID {
+	return c.uuid.UUID()
+}
+
+// DiscoverCharacteristics discovers characteristics in this service. Pass a
+// list of characteristic UUIDs you are interested in to this function. Either a
+// list of all requested services is returned, or if some services could not be
+// discovered an error is returned. If there is no error, the characteristics
+// slice has the same length as the UUID slice with characteristics in the same
+// order in the slice as in the requested UUID list.
+//
+// Passing a nil slice of UUIDs will return a complete
+// list of characteristics.
+func (s *DeviceService) DiscoverCharacteristics(uuids []UUID) ([]DeviceCharacteristic, error) {
+	return nil, nil
+}
```

Comment:
I guess this is simply not implemented yet?

To make this mergeable, I'd like this to either not exist or return an error ("not implemented" or something like that), so that people won't be confused why the method doesn't work.

The same goes for a few other methods later in this file.

Reply:
added "not yet implemented" error here.